### PR TITLE
feat: add SETUID and SETGID capabilities for gitlab runner container security context

### PR DESCRIPTION
**Pull request by ericwyles**
## Description

Add SETUID capability for gitlab runner container security context to allow running buildah inside a container from the runners.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-package-gitlab-runner/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

**Review by Racer159 — CHANGES_REQUESTED** [review]
I think this change should be fine but can you add a docs note to `configuration.md` to mention needing this `sysctl` setting? https://github.com/defenseunicorns/uds-rke2-image-builder/pull/81/files#diff-8c6b38593d0267b1fcfe569ffd1ab064b64bdecc6293bc967cded5f4b9c8bc13R15
(ideally in an admonition to call attention to it)

File: docs/configuration.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
 
 GitLab Runners in this package are configured through the upstream [GitLab Runner chart](https://docs.gitlab.com/runner/install/kubernetes.html) as well as a UDS configuration chart that supports the following:
 
+## Node Configuration
+
+> [!IMPORTANT]
+> Any kubernetes node that will run GitLab Runner pods must set the sysctl `user.max_user_namespaces` value to nonzero. This is required to run builds inside Linux containers from the runner pods.
```
Comment (Racer159) [inline docs/configuration.md:8]:
```suggestion
> Any kubernetes node that will run GitLab Runner pods to use tooling like [Buildah](https://buildah.io/) must set sysctl `user.max_user_namespaces` to a nonzero value. This is required to run these container builds inside Linux containers from the runner pods.
```

**Review by Racer159 — CHANGES_REQUESTED** [review]
Small tweak since it is only applicable to buildah and tools like it

**Comment by bburky** [comment]
CAP_SETUID is basically equivalent to root in the container. I understand there's a need for this for container builds, but I don't love seeing this be the default config for GitLab.

Do we support multiple configurations of gitlab runner groups (so this could be an optional config)? Do we support/suggest using dedicated nodes for gitlab runners (so that compromising CI doesn't compromise the whole cluster)?

Have we made any progress on using ephemeral EC2/VM-based runners? Let people have root in their CI (and use real `docker build`), if you've isolated them with a VM or dedicated EC2 instance.

**Comment by bburky** [comment]
I guess DISABLE_RESTRICTCAPABILITIES_EXEMPTION is disabled by default. But it does end up enabled for all runners.

**Comment by Racer159** [comment]
> CAP_SETUID is basically equivalent to root in the container. I understand there's a need for this for container builds, but I don't love seeing this be the default config for GitLab.
> 
> Do we support multiple configurations of gitlab runner groups (so this could be an optional config)? Do we support/suggest using dedicated nodes for gitlab runners (so that compromising CI doesn't compromise the whole cluster)?
> 
> Have we made any progress on using ephemeral EC2/VM-based runners? Let people have root in their CI (and use real `docker build`), if you've isolated them with a VM or dedicated EC2 instance.

We do support running gitlab runners on their own cluster (you can give it a different gitlab URL and registration token) and it is an optional config (though defaulted) as written - we can undefault it or remove that part of the PR to only partially support Buildah in the runner (buildah can run without this but flags warnings and I would expect some container build operations would fail though haven't dug that far yet).

Not yet on fleeting runners though it is something still on the docket to setup for envs that can use it (really just AWS rn)

**Comment by Racer159** [comment]
Opened a draft PR for the fleeting plugin: https://github.com/defenseunicorns/uds-package-gitlab-runner/pull/117 - looks like it is out of beta now too cc @oates / @zachariahmiller

**Comment by zachariahmiller** [comment]
> > CAP_SETUID is basically equivalent to root in the container. I understand there's a need for this for container builds, but I don't love seeing this be the default config for GitLab.
> > Do we support multiple configurations of gitlab runner groups (so this could be an optional config)? Do we support/suggest using dedicated nodes for gitlab runners (so that compromising CI doesn't compromise the whole cluster)?
> > Have we made any progress on using ephemeral EC2/VM-based runners? Let people have root in their CI (and use real `docker build`), if you've isolated them with a VM or dedicated EC2 instance.
> 
> We do support running gitlab runners on their own cluster (you can give it a different gitlab URL and registration token) and it is an optional config (though defaulted) as written - we can undefault it or remove that part of the PR to only partially support Buildah in the runner (buildah can run without this but flags warnings and I would expect some container build operations would fail though haven't dug that far yet).
> 
> Not yet on fleeting runners though it is something still on the docket to setup for envs that can use it (really just AWS rn)

Unless something has drastically changed recently it fails without SETUID and honestly i thought SETGUID in rootless. IIRC its on the mount, `buildah bud` and `buildah unshare`

File: chart/templates/uds-policy-exemptions.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+spec:
+  exemptions:
+    - description: Allow more capabilities for container build tools (Buildah) to be able to map user and group IDs
+      policies:
+        - RestrictCapabilities
+      title: "gitlab-runner-container-building"
+      matcher:
+        namespace: gitlab-runner-sandbox
+        name: "^runner-.*"
+{{- end }}
```
Comment (zachariahmiller) [inline chart/templates/uds-policy-exemptions.yaml:16]: newline

File: tests/journey/pipeline-run.test.ts [hunk 3]
```diff
@@ -42,29 +98,39 @@ test('test kicking off a pipeline run', async () => {
...
+    });
     expect(runnerResp.status).toBe(200)
+}
 
-    // Check that the pipeline actually ran successfully
-    let foundTheKitteh = await retry(async () => {
-        const jobIDResp = await (await fetch(`https://gitlab.uds.dev/api/v4/projects/1/jobs`, { headers })).json()
+async function checkJobResults(projectId: any, headers: HeadersInit, expectedJobLogOutputs: string[], expectedStatus: string) {
+    let status = await retry(async () => {
+        const jobIDResp = await (await fetch(`https://gitlab.uds.dev/api/v4/projects/${projectId}/jobs`, { headers })).json()
```
Comment (zachariahmiller) [inline tests/journey/pipeline-run.test.ts:107]: We should probably pull out domain as a var as it was desired to have these tests be something that could be run from somewhere other than our CI in the future and domain would need to be provided.

File: tests/journey/pipeline-run.test.ts [hunk 4]
```diff
@@ -1,10 +1,62 @@
...
+import { rm } from 'fs/promises';
+
+const domainSuffix = process.env.DOMAIN_SUFFIX || ".uds.dev"
+
+test('hello kitteh succeeds', async () => {
+    const sourceRepoName = 'kitteh'
+    const expectedStatus = 'success'
+    const expectedJobLogOutputs: string[] = ['Hello Kitteh']
+
+    //await executeTest(sourceRepoName, expectedJobLogOutputs, expectedStatus)
```
Comment (Racer159) [inline tests/journey/pipeline-run.test.ts:15]: commented code?

**Review by Racer159 — CHANGES_REQUESTED** [review]

**Review by Racer159 — APPROVED** [review]
lgtm assuming tests pass

File: tests/journey/pipeline-run.test.ts (hunk 3 above)
Comment (ericwyles) [reply tests/journey/pipeline-run.test.ts]: pulled this out to an env var!

**Pull request merged**